### PR TITLE
Cannot Count(Expr:Wildcard) with DataFrame API

### LINKED ISSUE #5473: Cannot Count(Expr:Wildcard) with DataFrame API
**Describe the bug**

Cannot do `count(Expr::Wildcard)` aggregate with DataFrame API

**To Reproduce**

```rust
    ctx.table("alltypes_plain")
        .await?
        .aggregate(vec![], vec![count(Expr::Wildcard)])?
        .explain(false, false)?
        .show()
        .await?;
```

Throws error:

`Error: Internal("Wildcard expressions are not valid in a logical query plan")`

**Expected behavior**

Should be like when doing SQL:

```rust
    ctx.sql("select count(*) from alltypes_plain")
        .await?
        .explain(false, false)?
        .show()
        .await?;
```

Gives plan:

```
+---------------+---------------------------------------------------+
| plan_type     | plan                                              |
+---------------+---------------------------------------------------+
| logical_plan  | Aggregate: groupBy=[[]], aggr=[[COUNT(UInt8(1))]] |
|               |   TableScan: alltypes_plain projection=[id]       |
| physical_plan | ProjectionExec: expr=[7300 as COUNT(UInt8(1))]    |
|               |   EmptyExec: produce_one_row=true                 |
|               |                                                   |
+---------------+---------------------------------------------------+
```

Where `count(*)` has been transformed to `count(1)` essentially.

**Additional context**

Expansion substitute:

https://github.com/apache/arrow-datafusion/blob/e9852074bacd8c891d84eba38b3417aa16a2d18c/datafusion/expr/src/utils.rs#L43-L45

See SQL substitute function:

https://github.com/apache/arrow-datafusion/blob/e9852074bacd8c891d84eba38b3417aa16a2d18c/datafusion/sql/src/expr/function.rs#L193-L218

### PR DESCRIPTION
**Pull request by jiangzhx**
# Which issue does this PR close?
Closes #5473

### CODE REVIEW

**Comment by mingmwang** [comment]
I think such logic should be moved to the Analyzer.

**Comment by jiangzhx** [comment]
> 
thanks @mingmwang ,  as @Jefffrey mention 
SQL substitute function is here:
https://github.com/apache/arrow-datafusion/blob/e9852074bacd8c891d84eba38b3417aa16a2d18c/datafusion/sql/src/expr/function.rs#L193-L218

but  use dafaframe  will got error
```
    ctx.sql("select count(*) from alltypes_plain")
        .await?
        .explain(false, false)?
        .show()
        .await?;

```

**Comment by jiangzhx** [comment]
As per everyone's suggestions, I made some modifications. 
Can @mingmwang  and @Jefffrey  take a look and help with code review when you have time.

File: datafusion/expr/src/logical_plan/builder.rs
Comment (Jefffrey) [inline datafusion/expr/src/logical_plan/builder.rs:1035]: the extreme nesting is not ideal, i think can simplify to this:

```suggestion
//handle Count(Expr:Wildcard) with DataFrame API
pub fn handle_wildcard(exprs: Vec<Expr>) -> Result<Vec<Expr>> {
    let exprs: Vec<Expr> = exprs
        .iter()
        .map(|expr| match expr {
            Expr::AggregateFunction(AggregateFunction {
                fun: aggregate_function::AggregateFunction::Count,
                args,
                distinct,
                filter,
            }) if args.len() == 1 => match args[0] {
                Expr::Wildcard => Expr::AggregateFunction(AggregateFunction {
                    fun: aggregate_function::AggregateFunction::Count,
                    // TODO: replace with the constant
                    args: vec![lit(ScalarValue::UInt8(Some(1)))],
                    distinct: *distinct,
                    filter: filter.clone(),
                }),
                _ => expr.clone(),
            },
            _ => expr.clone(),
        })
        .collect();
    Ok(exprs)
}
```

unsure if can simplify more, feel free to explore that

also check my TODO comment on the constant (i think i mention it in the original issue)

**Comment by jiangzhx** [comment]
> I think such logic should be moved to the Analyzer.

@mingmwang 
I understand your architecture design now when I see your PR #5570. After the PR is merged, I will migrate the current logic into your solution.

**Comment by jiangzhx** [comment]
try adding AnalyzerRule #5627 to resolve this bug

**Review by Jefffrey — APPROVED** [review]
does look good, though worth checking other opinions about where the logic should reside (whether like this or in analyzer as mentioned)

File: datafusion/core/tests/dataframe.rs
Comment (alamb) [inline datafusion/core/tests/dataframe.rs:68]: 👍

**Review by alamb — APPROVED** [review]
LGTM -- I also think https://github.com/apache/arrow-datafusion/pull/5627 looks good to but defer to @mingmwang

**Comment by alamb** [comment]
Closing in favor of https://github.com/apache/arrow-datafusion/pull/5518 (so it is clear which we are trying to merge)

Thank you @jiangzhx  and @Jefffrey

**Pull request closed without merging**
